javaagent-tooling/src/testPatchBytecodeVersion/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/OldBytecode.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.tooling.instrumentation.indy;
+
+import java.lang.reflect.Modifier;
+import net.bytebuddy.ByteBuddy;
+import net.bytebuddy.ClassFileVersion;
+import net.bytebuddy.asm.AsmVisitorWrapper;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.dynamic.DynamicType;
+import net.bytebuddy.dynamic.scaffold.InstrumentedType;
+import net.bytebuddy.implementation.Implementation;
+import net.bytebuddy.implementation.bytecode.ByteCodeAppender;
+import org.objectweb.asm.Label;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+
+public class OldBytecode {
+
+  private OldBytecode() {}
+
+  /**
+   * Generates and run a simple class with a {@link #toString()} implementation as if it had been
+   * compiled on an older java compiler
+   *
+   * @param className class name
+   * @param version bytecode version
+   * @return "toString"
+   */
+  public static String generateAndRun(String className, ClassFileVersion version) {
+    try (DynamicType.Unloaded<Object> unloadedClass = makeClass(className, version)) {
+      Class<?> generatedClass = unloadedClass.load(OldBytecode.class.getClassLoader()).getLoaded();
+
+      return generatedClass.getConstructor().newInstance().toString();
+
+    } catch (Throwable e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static DynamicType.Unloaded<Object> makeClass(
+      String className, ClassFileVersion version) {
+    return new ByteBuddy(version)
+        .subclass(Object.class)
+        // required otherwise stack frames aren't computed when needed
+        .visit(
+            version.isAtLeast(ClassFileVersion.JAVA_V7)
+                ? new ComputeFramesAsmVisitorWrapper()
+                : AsmVisitorWrapper.NoOp.INSTANCE)
+        .name(className)
+        .defineMethod("toString", String.class, Modifier.PUBLIC)
+        .intercept(new ToStringMethod())
+        .make();
+  }
+
+  private static class ToStringMethod implements Implementation, ByteCodeAppender {
+
+    @Override
+    public ByteCodeAppender appender(Target implementationTarget) {
+      return this;
+    }
+
+    @Override
+    public InstrumentedType prepare(InstrumentedType instrumentedType) {
+      return instrumentedType;
+    }
+
+    @Override
+    public Size apply(
+        MethodVisitor methodVisitor,
+        Context implementationContext,
+        MethodDescription instrumentedMethod) {
+      boolean useJsrRet =
+          implementationContext.getClassFileVersion().isLessThan(ClassFileVersion.JAVA_V5);
```

Comment:
Any reason why `jsr` is used only for java 4 and older? I thought it would be valid to have it in java 6 and older class files (unless java 6 class file also has frames).

Reply:
Yes, from what I understand, the JSR/RET opcodes aren't expected for bytecode that is compiled for java 5 and later, here is a good explaination from the JVM specs: https://stackoverflow.com/questions/37013761/status-of-jsr-ret-in-jvm-spec.

Reply:
can you add a comment to the code since this isn't so obvious? thx